compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3746,17 +3747,22 @@ private BType getIntersectionForErrorTypes(IntersectionContext intersectionConte
     }
 
     private BType createRecordIntersection(IntersectionContext diagnosticContext,
-                                           BType recordTypeOne, BType recordTypeTwo, SymbolEnv env) {
+                                           BRecordType recordTypeOne, BRecordType recordTypeTwo, SymbolEnv env) {
 
         BRecordType newType = createAnonymousRecord(env);
 
-        if (!populateRecordFields(diagnosticContext.switchLeft(), newType, recordTypeOne, env, null) ||
-                !populateRecordFields(diagnosticContext.switchRight(), newType, recordTypeTwo, env, null)) {
+        BType restFieldTypeOne = recordTypeOne.restFieldType;
+        BType restFieldTypeTwo = recordTypeTwo.restFieldType;
+
+        BType constraintOne = restFieldTypeOne == null || restFieldTypeOne == symTable.noType ? null : restFieldTypeOne;
+        BType constraintTwo = restFieldTypeTwo == null || restFieldTypeTwo == symTable.noType ? null : restFieldTypeTwo;
```

Comment:
Are these null checks necessary? IIRC, if there's no rest field, we set the `symTable.noType` as the rest field type.

Reply:
I was wondering if we could get to this method when the record type is still being defined and `noType` hasn't been set yet. But this check is redundant anyway since even if it was null the else logic will set null as the constraint. Fixed in a26b746, thanks!

Reply:
Had to change this logic a bit to properly handle closed records. Now the never type is used as the constraint if the record is a closed record.